kernel/posix/pthread.c

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <atomic.h>
+#include "ksched.h"
+#include "wait_q.h"
+
+/* Pthread structure for number of threads */
+struct posix_thread pthread[CONFIG_MAX_PTHREAD_COUNT];
+
+static bool is_posix_prio_valid(u32_t priority, int policy)
+{
+	if (priority >= sched_get_priority_min(policy) &&
+	     priority <= sched_get_priority_max(policy)) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline u32_t zephyr_to_posix_priority(s32_t z_prio, int *policy)
+{
+	u32_t prio;
+
+	if (z_prio < 0) {
+		*policy = SCHED_FIFO;
+		prio = -1 * (z_prio + 1);
+	} else {
+		*policy = SCHED_RR;
+		prio = (CONFIG_NUM_PREEMPT_PRIORITIES - z_prio);
+	}
+
+	return prio;
+}
+
+static inline s32_t posix_to_zephyr_priority(u32_t priority, int policy)
+{
+	s32_t prio;
+
+	if (policy == SCHED_FIFO) {
+		/* Zephyr COOP priority starts from -1 */
+		prio =  -1 * (priority + 1);
+	} else {
+		prio = (CONFIG_NUM_PREEMPT_PRIORITIES - priority);
+	}
+
+	return prio;
+}
+
+int pthread_attr_setschedparam(pthread_attr_t *attr,
+			       const struct sched_param *schedparam)
+{
+	int priority = schedparam->priority;
+
+	if (is_posix_prio_valid(priority, attr->schedpolicy) == false) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	attr->priority = priority;
+	return 0;
+}
+
+int pthread_attr_setstack(pthread_attr_t *attr, void *stackaddr,
+			  size_t stacksize)
+{
+	if (stackaddr == NULL) {
+		errno = EACCES;
+		return -1;
+	}
+
+	attr->stack = stackaddr;
+	attr->stacksize = stacksize;
+	return 0;
+}
+
+static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+{
+	void * (*fun_ptr)(void *);
+
+	fun_ptr = arg3;
+	(*fun_ptr)((void *)arg1);
+}
+
+int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
+		   void *(*threadroutine)(void *), void *arg)
+{
+	static atomic_val_t thread_count;
+	atomic_val_t thread_index = 0;
+	s32_t prio;
+	pthread_mutexattr_t att;
+	pthread_condattr_t cond_attr;
+
+	if (!attr || !attr->stack) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	prio = posix_to_zephyr_priority(attr->priority, attr->schedpolicy);
+	thread_index = atomic_inc(&thread_count);
+
+	if (thread_index < CONFIG_MAX_PTHREAD_COUNT) {
+		pthread[thread_index].cancel_state = (1 << PTHREAD_CANCEL_POS) &
```

Comment:
It's cleaner to just get a pointer to `pthread[thread_index]` and set `thread->cancel_state`, `thread->state`, `thread->cancel_pending`, etc.

Might want to add a `FIXME` comment saying that this will change to a mempool (although it's, IMO, faster to just write it as a mempool right away; it's simple and will fix the issue of running out of `pthread` handles to give out if threads are created and destroyed).

Reply:
Addressed.